wfcGen.py

```diff
@@ -164,85 +166,97 @@ def __init__(self, sx, sy, cells, tileSize=3):
-                    self.entropySave[entropy] = self.entropySave.get(entropy, []) + [[x, y]]
+        rand = [random.randint(0, self.sy-1), random.randint(0, self.sx-1)]
+        self.grid[rand[0]][rand[1]].setCell(random.randint(0, len(self.cells)-1), self.cells)
+        self.loadPosEntropy([rand[1], rand[0]])
+        self.loadEntropyAroundPos([rand[1], rand[0]])
+
+    def loadPosEntropy(self, pos):
+        if (pos[0], pos[1]) in list(self.invertedEntropy.keys()) and pos in self.entropySave[self.invertedEntropy[(pos[0], pos[1])]]:
+            self.entropySave[self.invertedEntropy[(pos[0], pos[1])]].remove(pos)
+            if self.entropySave[self.invertedEntropy[(pos[0], pos[1])]] == []:
+                self.entropySave.pop(self.invertedEntropy[(pos[0], pos[1])], None)
+        self.grid[pos[1]][pos[0]].loadEntropy(self.grid)
+        entropy = self.grid[pos[1]][pos[0]].getEntropy(self.cells)
+        if entropy != -1:
+            self.entropySave[entropy] = self.entropySave.get(entropy, []) + [pos]
+            self.invertedEntropy[(pos[0], pos[1])] = entropy
+        else:
+            self.invertedEntropy.pop((pos[0], pos[1]), None)
+    
+    def loadEntropyAroundPos(self, pos):
+        for x in [-1, 0, 1]:
+            for y in [-1, 0, 1]:
+                if abs(x) == abs(y):
+                    continue
+                if (pos[1]+y)<len(self.grid) and (pos[0]+x)<len(self.grid) and (pos[1]+y)>=0 and (pos[0]+x)>=0:
+                    self.loadPosEntropy([pos[0]+x, pos[1]+y])
 
     def collapse(self):
-        self.loadEntropy()
         sortedEntropy = sorted(self.entropySave.keys())
         if len(sortedEntropy) == 0:
                 return True
         while min(sortedEntropy) < 1:
             sortedEntropy.remove(min(sortedEntropy))
             if len(sortedEntropy) == 0:
                 return True
-        y = self.entropySave[sortedEntropy[0]][0][1]
-        x = self.entropySave[sortedEntropy[0]][0][0]
+        entropyPos = random.choice(self.entropySave[sortedEntropy[0]])
+        y = entropyPos[1]
+        x = entropyPos[0]
 
         self.grid[y][x].collapse(self.cells)
-        self.loadEntropy()
-        ZeroCheck = 0 in self.entropySave.keys()
-        while 0 in self.entropySave.keys():
-            if len(self.backUps) > 0:
-                lb = self.backUps[len(self.backUps)-1]
-                self.grid[lb[1]][lb[0]] = Cell(lb)
-                self.backUps = self.backUps[:-1]
-            else:
-                self.setRandom()
-            self.loadEntropy()
-        if ZeroCheck:
+        self.loadPosEntropy([x, y])
+        self.loadEntropyAroundPos([x, y])
+        if 0 in self.entropySave:
             self.reloads += 2
-            for j in range(self.reloads + random.randint(0, 4)):
+            for i in range(self.reloads + random.randint(0, 3)):
                 if len(self.backUps) > 0:
                     lb = self.backUps[len(self.backUps)-1]

```

Comment:
Should this be unpacked too?

Reply:
what unpacked?
I don't see anything there to unpack

Reply:
I mean `lb`. Have a look at how you use `lb` later. It seems like you mostly need the two components of it, rather than the whole tuple